### PR TITLE
Rescheduler for dspods only

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
This PR:
- Ensures rescheduler is run for pods which are created from DS controller.
- Rebases kube to 1.10 and bumps Godeps.

/cc @bsalamat

### CODE REVIEW

File: rescheduler/rescheduler_test.go [hunk 1]
```diff
@@ -52,24 +52,25 @@ func TestWaitForScheduled(t *testing.T) {
...
 	assert.Equal(t, 0, len(filtered))
 
-	allPods = []*apiv1.Pod{
-		createTestPod("heapster", true, 0),
-		createTestPod("random", false, 0),
-		createTestPod("dns", true, 0),
+	allPods = []*v1.Pod{
+		createTestPod("heapster", true, true, 0),
+		createTestPod("random1", false, false, 0),
+		createTestPod("random1", true, false, 0), // Eventhough this is criticalPod, this is not created by DS.
```
Comment (ravisantoshgudimetla) [inline rescheduler/rescheduler_test.go]: This explains the crux of changes.

File: rescheduler/rescheduler.go [hunk 2]
```diff
@@ -371,23 +375,19 @@ func prepareNodeForPod(client kube_client.Interface, recorder kube_record.EventR
...
 }
 
-func copyNode(node *apiv1.Node) (*apiv1.Node, error) {
-	objCopy, err := api.Scheme.DeepCopy(node)
-	if err != nil {
-		return nil, err
-	}
-	copied, ok := objCopy.(*apiv1.Node)
+func copyNode(obj interface{}) (*v1.Node, error) {
+	copied, ok := obj.(*v1.Node)
```
Comment (bsalamat) [inline rescheduler/rescheduler.go]: Why did you remove DeepCopy? The object is not copied anymore.

File: rescheduler/rescheduler.go [hunk 3]
```diff
@@ -458,17 +458,39 @@ func groupPods(client kube_client.Interface, node *apiv1.Node) ([]*apiv1.Pod, []
...
 		}
 	}
 	return criticalPods
 }
 
-func isCriticalPod(pod *apiv1.Pod) bool {
+func isCriticalPod(pod *v1.Pod) bool {
+	// Shouldn't we check "kube-system" namespace as well?
 	_, found := pod.ObjectMeta.Annotations[criticalPodAnnotation]
+	criticalBasedOnPriority := false
```
Comment (bsalamat) [inline rescheduler/rescheduler.go]: If `found` is true, you can return early:

```go
func isCriticalPod(pod *v1.Pod) bool {
	if _, found := pod.ObjectMeta.Annotations[criticalPodAnnotation]; found {
		return true
	}
	return pod.Spec.Priority != nil && *pod.Spec.Priority >= SystemCriticalPriority
}
```

File: rescheduler/rescheduler.go [hunk 4]
```diff
@@ -146,10 +150,10 @@ func main() {
 					continue
 				}
 
-				criticalPods := filterCriticalPods(allUnschedulablePods, podsBeingProcessed)
+				criticalPodsCreatedByDaemonSets := filterCriticalPodsCreatedByDaemonSets(allUnschedulablePods, podsBeingProcessed)
```
Comment (bsalamat) [inline rescheduler/rescheduler.go]: optional: s/criticalPodsCreatedByDaemonSets/criticalDaemonSetPods/ to be shorter

**Review by bsalamat — COMMENTED** [review]
Thanks, @ravisantoshgudimetla! Looks generally good. I have several minor comments.

File: rescheduler/rescheduler.go (hunk 2 above)
Comment (ravisantoshgudimetla) [reply rescheduler/rescheduler.go]: Yes, you are right. I missed that part. I wanted to do a `node.DeepCopy()` and remove this function.

File: rescheduler/rescheduler.go (hunk 3 above)
Comment (ravisantoshgudimetla) [reply rescheduler/rescheduler.go]: Yes, thanks. I will make that change.

File: rescheduler/rescheduler.go [hunk 5]
```diff
@@ -458,17 +458,39 @@ func groupPods(client kube_client.Interface, node *apiv1.Node) ([]*apiv1.Pod, []
...
+		if isCriticalPod(pod) && isDaemonsetPod(pod) && !podsBeingProcessed.Has(pod) {
 			criticalPods = append(criticalPods, pod)
 		}
 	}
 	return criticalPods
 }
 
-func isCriticalPod(pod *apiv1.Pod) bool {
+func isCriticalPod(pod *v1.Pod) bool {
+	// Shouldn't we check "kube-system" namespace as well?
```
Comment (ravisantoshgudimetla) [inline rescheduler/rescheduler.go]: What about this comment?

**Comment by ravisantoshgudimetla** [comment]
Thanks @bsalamat for the review. I have made modifications as per suggested changes. PTAL.

File: rescheduler/rescheduler.go [hunk 6]
```diff
@@ -458,17 +448,54 @@ func groupPods(client kube_client.Interface, node *apiv1.Node) ([]*apiv1.Pod, []
...
+	val, ok := annotations[criticalPodAnnotation]
+	if ok && val == "" {
+		return true
+	}
+	return false
+}
+
+// isCriticalPodBasedOnPriority checks if the given pod is a critical pod based on priority resolved from pod Spec.
+func isCriticalPodBasedOnPriority(priority int32) bool {
+	if priority >= SystemCriticalPriority {
```
Comment (bsalamat) [inline rescheduler/rescheduler.go]: nit: you can simply replace the whole body of the function with:
return priority >= SystemCriticalPriority

**Review by bsalamat — COMMENTED** [review]
Just a minor nit

File: rescheduler/rescheduler.go (hunk 6 above)
Comment (ravisantoshgudimetla) [reply rescheduler/rescheduler.go]: Yup, copy paste mistake for kube codebase!

**Comment by bsalamat** [comment]
@piosz Is the latest version of rescheduler automatically picked up and built with the rest of Kubernetes?
